reference/src/glossary.md

```diff
@@ -56,6 +55,24 @@ somewhat differently from this definition. However, that's considered a low
 level detail of a particular Rust implementation. When programming Rust, the
 Abstract Rust Machine is intended to operate according to the definition here.
 
+#### (Pointer) Provenance
+
+The *provenance* of a pointer can be used to distinguish pointers that point to the same memory location.
```

Comment:
No, I meant location -- as in, memory address. But it seems the term "location" is a very confusing one, which surprises me...

Reply:
I thought provenance could be used to distinguish pointers that point to the same memory allocation, not necessarily exclusively to the same memory location. 

For example, if I allocate an array, I can derive pointers to each array element, from the pointer to the array, and I can compare them, not only for equality, but also whether their addresses are larger or smaller than one another, etc. 

I cannot compare any of those pointers with pointers to other memory allocations.

Reply:
Are we using the same terms here?

"allocation" is a "block of memory", it's the thing created by `malloc`.
"location" is basically an alias for "memory address".

When considering something like
```rust
let raw1 = Box::into_raw(Box::new(0u16)) as *mut u8;
let raw2 = raw1.wrapping_add(1);
assert_ne!(raw1, raw2);
```
then both `raw1` and `raw2` point to the same *allocation*, but they point to different *locations*, making them easy to distinguish as demonstrated by the `assert_ne!`. In contrast, the two pointers in my example in the PR point to the same *location*, so they are not so easily distinguished, but they are still different.

Reply:
Maybe the reason you are confused because in C (ignoring `restrict`), the provenance of a pointer is basically an identifier for the allocation it originally pointed to. So in C, pointers to the same *location* have to still be distinguioshed if their provenance indicates they came from a different *allocation*.

That's just one example of pointer provenance though. Both `restrict` and Stacked Borrows use more refined provenance.

Reply:
> Are we using the same terms here?

I think so.

> then both raw1 and raw2 point into the same allocation, but they point to different locations, making them easy to distinguish as demonstrated by the assert_ne!

My point is that provenance is what allows you to make this distinction. If both raw1 and raw2 where to point to different allocations, the behavior of the `assert_ne!` would be undefined (at least in C and C++ where comparing pointers to different allocations is illegal).

> Maybe the reason you are confused is that in C (ignoring restrict), the provenance of a pointer is basically an identifier for the allocation it originally pointed to.

Yep, this might be it.

Reply:
> My point is that provenance is what allows you to make this distinction.

Exactly. And that is what my text says, and your edited text does *not* say.

`raw2` and `raw2_wrong` *point to the same allocation*. I define "pointing to an allocation" to mean "look at the address the pointer points to, and check which allocation it lies in". So, by definition, pointers pointing to the same location also point to the same allocation, as "allocation" is a strictly more coarse categorization than location.

However, `raw2` and `raw2_wrong` *originate* from different allocations, and that is what the provenance records.

> So you can't use provenance to distinguish pointers to different allocations, only to different locations within the same allocation there.

Provenance is the only thing that distinguishes pointers that point to the same location but are still different.

Maybe what you mean is *the program* cannot use provenance, and that is true, the program does not have any access to provenance. The party distinguishing here is the spec writer / the abstract machine.

Reply:
Ahhh, so what you mean is that provenance allows you to tell whether two pointers to the same memory location come from the same or from different allocations? 

The reason I find that way to put it confusing, is that it sounds as if you could use provenance to actually write a function that returns true whether two pointers to the same location come from the same allocation, and false otherwise. But AFAICT this is not possible, since comparing pointers to different allocations in any way invokes undefined behavior.

Reply:
@gnzlbg I have expanded the example and tried to clarify the wording. Does this help?

Reply:
> comparing pointers to different allocations in any way invokes undefined behavior.

It certainly is not UB as you can do it in safe Rust.

> it sounds as if you could use provenance to actually write a function that returns true whether two pointers to the same location come from the same allocation

I am open to different wording, but your suggested wording is IMO factually wrong. ;)
Provenance is extra state that the Abstract Machine tracks but the program has no access to. The function you are suggesting *could* exist on the Abstract Machine. It doesn't, because it couldn't be implemented by compilers in an efficient way.

Reply:
I think now it is clearer.

> It certainly is not UB as you can do it in safe Rust.

I think we should document this somewhere, since this is UB in C and C++. You cannot compare pointers to different allocations at all. For example [C11.6.5.8.5](https://port70.net/~nsz/c/c11/n1570.html#6.5.8p5) says (emphasis mine):

> When two pointers are compared, the result depends on the relative locations in the address space of the objects pointed to. If two pointers to object types both point to the same object, or both point one past the last element of the same array object, they compare equal. If the objects pointed to are members of the same aggregate object, pointers to structure members declared later compare greater than pointers to members declared earlier in the structure, and pointers to array elements with larger subscript values compare greater than pointers to elements of the same array with lower subscript values. All pointers to members of the same union object compare equal. If the expression P points to an element of an array object and the expression Q points to the last element of the same array object, the pointer expression Q+1 compares greater than P. **In all other cases, the behavior is undefined.**

Reply:
I know it is UB in C/C++, though many programs do it anyway so UB likely shouldn't be as aggressively defined. But that is out of scope here.

Also note that this clause *does* allow comparing pointers from different allocations under some very specific conditions:

> If two pointers to object types both point to the same object, or both point one past the last element of the same array object, they compare equal.

---

> I think we should document this somewhere

What would be a good place to say things like "oh btw, did you know comparing raw pointers is safe in Rust"?

To be clear, having raw ptr comparison in safe Rust was a mistake IMO, and making sure ptr comparison is never UB will be "fun". Or rather, it actually is not UB in LLVM ever, but it *can* return `undef`/`poison` (at least in [the latest formal model of that part of LLVM](https://people.mpi-sws.org/~jung/twinsem/twinsem.pdf)), which is just as bad to have in safe Rust.

Reply:
Anyway, marking as resolved based on "I think now it is clearer".

Reply:
> Also aren't you allowed to compare the one-past-the-end ptr of one object with the first ptr of an adjacent object, or so?

Yep, that's the second sentence in the quoted paragraph. The provenance of the pointer to one past the end of an object must be, however, the same as that of the object. If you perform two independent allocations, and the second one happens to be adjacent, it does not make a pointer to it a "one past the end" pointer of the first allocation.

> What would be a good place to say things like "oh btw, did you know comparing raw pointers is safe in Rust"?

I think it would be better to phrase this information as "oh btw, provenance in Rust is subtly difference from provenance in C and C++ in these particular ways: ... * comparing raw pointers to different allocations is ok in Rust".

Reply:
The provenance isn't different. Just the way in which provenance is declared to (not) cause UB is.

Reply:
Maybe the nomicon would be a better place to call this out.

Reply:
Maybe. The details of provenance are not widely known, so if we can start by getting consensus in the UCG WG that this is a reasonable definition, that'd already be great.

Reply:
The latest changes look good to me, thanks!